x/session/keeper/session_hydrator.go

```diff
@@ -0,0 +1,203 @@
+	appAddress string,
+	serviceId string,
+	blockHeight int64,
+) *sessionHydrator {
+	sessionHeader := &types.SessionHeader{
+		ApplicationAddress: appAddress,
+		ServiceId:          &sharedtypes.ServiceId{Id: serviceId},
+	}
+	return &sessionHydrator{
+		sessionHeader: sessionHeader,
+		session:       &types.Session{},
+		blockHeight:   blockHeight,
+		sessionIdBz:   make([]byte, 0),
+	}
+}
+
+// GetSession implements of the exposed `UtilityModule.GetSession` function
+// TECHDEBT(#519): Add custom error types depending on the type of issue that occurred and assert on them in the unit tests.
+func (k Keeper) HydrateSession(ctx sdk.Context, sh *sessionHydrator) (*types.Session, error) {
+	logger := k.Logger(ctx).With("method", "hydrateSession")
+
+	if err := k.hydrateSessionMetadata(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate the session metadata: %v", err)
+	}
+	logger.Debug("Finished hydrating session metadata")
+
+	if err := k.hydrateSessionID(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate the session ID: %v", err)
+	}
+	logger.Info("Finished hydrating session ID: %s", sh.sessionHeader.SessionId)
+
+	if err := k.hydrateSessionApplication(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate application for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session application: %+v", sh.session.Application)
+
+	if err := k.hydrateSessionSuppliers(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate suppliers for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session suppliers: %+v")
+
+	sh.session.Header = sh.sessionHeader
+	sh.session.SessionId = sh.sessionHeader.SessionId
+
+	return sh.session, nil
+}
+
+// hydrateSessionMetadata hydrates metadata related to the session such as the height at which the session started, its number, the number of blocks per session, etc..
+func (k Keeper) hydrateSessionMetadata(ctx sdk.Context, sh *sessionHydrator) error {
+	sh.session.NumBlocksPerSession = NumBlocksPerSession
+	sh.session.SessionNumber = int64(sh.blockHeight / NumBlocksPerSession)
+	sh.sessionHeader.SessionStartBlockHeight = sh.blockHeight - (sh.blockHeight % NumBlocksPerSession)
+	return nil
+}
+
+// hydrateSessionID use both session and on-chain data to determine a unique session ID
+func (k Keeper) hydrateSessionID(ctx sdk.Context, sh *sessionHydrator) error {
+	// TODO_TECHDEBT: Need to retrieve the block hash at SessionStartBlockHeight, NOT THE CURRENT ONE
+	prevHashBz := ctx.HeaderHash()
+	appPubKeyBz := []byte(sh.sessionHeader.ApplicationAddress)
+	serviceIdBz := []byte(sh.sessionHeader.ServiceId.Id)
+	sessionHeightBz := make([]byte, 8)
+	binary.LittleEndian.PutUint64(sessionHeightBz, uint64(sh.sessionHeader.SessionStartBlockHeight))
+
+	sh.sessionIdBz = concat(prevHashBz, serviceIdBz, appPubKeyBz, sessionHeightBz)
```

Comment:
Since `serviceIdBz` may be of a variable length, we should either pad it or have some kind of separator in between the `concat` parameters. Something like:

```suggestion
        separator := []byte("some-constant-separator")
	sh.sessionIdBz = concat(prevHashBz, separator, serviceIdBz, separator, appPubKeyBz, separator, sessionHeightBz)
```